### PR TITLE
Fix crash caused by data race in useAnimatedKeyboard

### PR DESCRIPTION
**Pull request by graszka22**
## Summary

Fixes https://github.com/software-mansion/react-native-reanimated/issues/3568
Fixes https://github.com/software-mansion/react-native-reanimated/issues/3712
Related to https://github.com/software-mansion/react-native-reanimated/pull/3704

The crash was caused because `_listeners` could be accessed by two threads at the same time. For example it could happen that while js thread in `unsubscribeFromKeyboardEvents` was deleting a listener, `updateKeyboardFrame` was iterating over it on the main thread. Now it's synchronised, so only one thread can access `_listeners` at the same time (and `_state` too just in case)

## Test plan

You can reproduce the crash easier by mounting 1000 components that use `useAnimatedKeyboard` in AnimatedKeyboardExample (I tested on simulator). Just enter and exit the example a couple of times.
<details>
  <summary>Example</summary>
  
  ```js
import Animated, {
  useAnimatedStyle,
  useAnimatedKeyboard,
  KeyboardState,
} from 'react-native-reanimated';
import {
  View,
  Button,
  TextInput,
  StyleSheet,
  Keyboard,
  ScrollView,
} from 'react-native';
import React from 'react';

const BOX_SIZE = 50;

function NestedView(): React.ReactElement {
  useAnimatedKeyboard();
  return <View style={styles.nestedView} />;
}

function AnimatedStyleUpdateExample(): React.ReactElement {
  const keyboard = useAnimatedKeyboard();
  const OPENING = KeyboardState.OPENING;
  const style = useAnimatedStyle(() => {
    const color = keyboard.state.value === OPENING ? 'red' : 'blue';

    return {
      backgroundColor: color,
    };
  });
  const translateStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateY: -keyboard.height.value }],
    };
  });
  const [shouldShowNestedView, setShouldShowNestedView] = React.useState(false);

  return (
    <ScrollView
      contentContainerStyle={styles.container}
      keyboardDismissMode="interactive"
      scrollEnabled={false}>
      <Animated.View style={[styles.box, style]} />
      <Button
        title="Toggle nested view"
        onPress={() => {
          setShouldShowNestedView(!shouldShowNestedView);
        }}
      />
      {new Array(1000).fill(null).map((_, i) => (
        <NestedView key={i} />
      ))}
      <Animated.View style={translateStyle}>
        <Button
          title="Dismiss"
          onPress={() => {
            Keyboard.dismiss();
          }}
        />
        <TextInput style={styles.textInput} autoCorrect  autoFocus/>
      </Animated.View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingVertical: 70,
  },
  box: { width: BOX_SIZE, height: BOX_SIZE, marginBottom: 100 },
  textInput: {
    borderColor: 'blue',
    borderStyle: 'solid',
    borderWidth: 2,
    height: 60,
    width: 200,
  },
  nestedView: {
    width: 100,
    height: 100,
    backgroundColor: '#ffff00',
  },
});

export default AnimatedStyleUpdateExample;

  ```
</details>

### CODE REVIEW

**Comment by tomekzaw** [comment]
I'm just wondering if it would be possible to schedule a call on the UI thread from JS thread instead of using `@synchronize` everywhere. 

It seems like the only two methods that are called on the JS thread are `subscribeFromKeyboardEvents` and `unsubscribeFromKeyboardEvents`. The latter should be easy as it doesn't return anything. On the other hand, `subscribeFromKeyboardEvents` returns the listener id, but we can assign it while still on the JS thread and copy the integer into the block that will be executed on the UI thread.

This would mean that the registration of listeners is asynchronous, but we no longer need synchronization in notification methods (`keyboardDidShow` etc.).

**Comment by graszka22** [comment]
@tomekzaw  That works too :)

**Review by tomekzaw — APPROVED** [review]
Looks good to me. One thing that needs to be checked before merging is whether reloading still works because using `self` in blocks directly may lead to reference cycles.

**Comment by graszka22** [comment]
Reloading doesn't work anyway https://github.com/software-mansion/react-native-reanimated/issues/3786 and I'm currently trying to fix it. It crashes for animated sensors too and probably in other places (https://github.com/software-mansion/react-native-reanimated/issues/3895). We can hold off on merging this until reloading is fixed.

**Pull request merged**
